plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java

```diff
@@ -524,6 +536,55 @@ else if (prop.size() == 1) {
         }
     }
 
+    private ColumnMapping ipAddressColumnMapping(String writeBindExpression)
+    {
+        return ColumnMapping.sliceMapping(
+                ipAddressType,
+                (resultSet, columnIndex) -> {
+                    byte[] address = InetAddresses.forString(resultSet.getString(columnIndex)).getAddress();
+
+                    byte[] bytes;
+                    if (address.length == 4) {
+                        bytes = new byte[16];
+                        bytes[10] = (byte) 0xff;
+                        bytes[11] = (byte) 0xff;
+                        arraycopy(address, 0, bytes, 12, 4);
+                    }
+                    else if (address.length == 16) {
+                        bytes = address;
+                    }
+                    else {
+                        throw new TrinoException(GENERIC_INTERNAL_ERROR, "Invalid InetAddress length: " + address.length);
+                    }
+
+                    return wrappedBuffer(bytes);
+                },
+                ipAddressWriteFunction(writeBindExpression));
+    }
+
+    private static SliceWriteFunction ipAddressWriteFunction(String bindExpression)
+    {
+        return new SliceWriteFunction() {
+            @Override
+            public String getBindExpression()
+            {
+                return bindExpression;
+            }
+
+            @Override
+            public void set(PreparedStatement statement, int index, Slice value)
+                    throws SQLException
+            {
+                try {
+                    statement.setObject(index, InetAddresses.toAddrString(InetAddress.getByAddress(value.getBytes())), Types.OTHER);
+                }
+                catch (UnknownHostException e) {
```

Comment:
Today I learnt that some methods in InetAddress are non-blocking (i.e. do not perform remote lookups) while some are. And both variants can throw the same exceptions - good luck trying to easily identify if the method used is a blocking one or not without looking at JavaDocs.